pin/pin.go

```diff
@@ -276,8 +276,9 @@ func (p *pinner) isPinnedWithType(c *cid.Cid, mode PinMode) (string, bool, error
 	}
 
 	// Default is Indirect
+	visitedSet := cid.NewSet()
 	for _, rc := range p.recursePin.Keys() {
-		has, err := hasChild(p.dserv, rc, c)
+		has, err := hasChild(p.dserv, rc, c, visitedSet.Visit)
```

Comment:
I dont think we can reuse the same `visitedSet` across pin checks, unless we pass all the pins we're checking into the call at once. otherwise we might mark a subtree as visited that contains the next pin we're checking